wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineSourceDto.java

```diff
@@ -30,10 +31,33 @@
     MachineSourceDto withType(String type);
 
     @Override
-    @FactoryParameter(obligation = MANDATORY)
+    @FactoryParameter(obligation = OPTIONAL)
```

Comment:
I think I've added a condition in WorkspaceService but I might add a new check in Factory Validator

I've not found an annotation @FactoryParameter with a "either content, either location" entry
Like "mutually exclusive"